Tests/test_file_avif.py

```diff
@@ -0,0 +1,830 @@
+        assert _avif.encoder_codec_available("aom") is True
+
+    def test_encoder_codec_available_bad_params(self) -> None:
+        with pytest.raises(TypeError):
+            _avif.encoder_codec_available()
+
+    @skip_unless_avif_decoder("dav1d")
+    @skip_unless_feature("avif")
+    def test_encoder_codec_available_cannot_decode(self) -> None:
+        assert _avif.encoder_codec_available("dav1d") is False
+
+    def test_encoder_codec_available_invalid(self) -> None:
+        assert _avif.encoder_codec_available("foo") is False
+
+    def test_encoder_quality_valueerror(self, tmp_path: Path) -> None:
+        with Image.open("Tests/images/avif/hopper.avif") as im:
+            test_file = str(tmp_path / "temp.avif")
+            with pytest.raises(ValueError):
+                im.save(test_file, quality="invalid")
+
+    @skip_unless_avif_decoder("aom")
+    @skip_unless_feature("avif")
+    def test_decoder_codec_available(self) -> None:
+        assert _avif.decoder_codec_available("aom") is True
+
+    def test_decoder_codec_available_bad_params(self) -> None:
+        with pytest.raises(TypeError):
+            _avif.decoder_codec_available()
+
+    @skip_unless_avif_encoder("rav1e")
+    @skip_unless_feature("avif")
+    def test_decoder_codec_available_cannot_decode(self) -> None:
+        assert _avif.decoder_codec_available("rav1e") is False
+
+    def test_decoder_codec_available_invalid(self) -> None:
+        assert _avif.decoder_codec_available("foo") is False
+
+    @pytest.mark.parametrize("upsampling", ["fastest", "best", "nearest", "bilinear"])
+    def test_decoder_upsampling(self, upsampling: str) -> None:
+        AvifImagePlugin.CHROMA_UPSAMPLING = upsampling
+        try:
+            with Image.open(TEST_AVIF_FILE):
+                pass
+        finally:
+            AvifImagePlugin.CHROMA_UPSAMPLING = "auto"
+
+    def test_decoder_upsampling_invalid(self) -> None:
+        AvifImagePlugin.CHROMA_UPSAMPLING = "foo"
+        try:
+            with pytest.raises(ValueError):
+                with Image.open(TEST_AVIF_FILE):
+                    pass
+        finally:
+            AvifImagePlugin.CHROMA_UPSAMPLING = "auto"
+
+    def test_p_mode_transparency(self) -> None:
+        im = Image.new("P", size=(64, 64))
+        draw = ImageDraw.Draw(im)
+        draw.rectangle(xy=[(0, 0), (32, 32)], fill=255)
+        draw.rectangle(xy=[(32, 32), (64, 64)], fill=255)
+
+        buf_png = BytesIO()
+        im.save(buf_png, format="PNG", transparency=0)
+        im_png = Image.open(buf_png)
+        buf_out = BytesIO()
+        im_png.save(buf_out, format="AVIF", quality=100)
+
+        assert_image_similar(im_png.convert("RGBA"), Image.open(buf_out), 1)
+
+    def test_decoder_strict_flags(self) -> None:
+        # This would fail if full avif strictFlags were enabled
+        with Image.open("Tests/images/avif/chimera-missing-pixi.avif") as im:
+            assert im.size == (480, 270)
```

Comment:
Isn't it possible that someone could build libavif with those strict flags enabled, and then build Pillow from source to use it? So in that scenario, this test would fail?

Reply:
`strictFlags` are passed at runtime, they're not set at compile time. By default it strictly enforces various specification rules, but two in particular are often found in the wild and so are regularly disabled:

https://github.com/python-pillow/Pillow/blob/8b8bbba0f77c4ed5d1479d3f6e0c1b79dfb4f867/src/_avif.c#L797-L802

[Chromium](https://chromium.googlesource.com/chromium/src/+/1ade7e92edd967c0d54c64f262f9c149b48d0a96/third_party/blink/renderer/platform/image-decoders/avif/avif_image_decoder.cc#660) disables these two flags and [WebKit](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/platform/image-decoders/avif/AVIFImageReader.cpp#L43-L47) disables `AVIF_STRICT_PIXI_REQUIRED`.
The `test_decoder_strict_flags` test verifies that the strictFlags are set appropriately so that these somewhat common images can load without throwing an error.